In [23]:
# import sys
# !{sys.executable} -m pip install gensim

     |████████████████████████████████| 24.2MB 848kB/s eta 0:00:01   |███████████▏                    | 8.4MB 33kB/s eta 0:07:55
     |████████████████████████████████| 122kB 440kB/s eta 0:00:01
  Created wheel for smart-open: filename=smart_open-3.0.0-cp37-none-any.whl size=107098 sha256=67dd73775e8bad336f264c2ad5c739adbb4e687eb4c4fe6993fb3a5157aa3372
  Stored in directory: /Users/levsolovev/Library/Caches/pip/wheels/18/88/7c/f06dabd5e9cabe02d2269167bcacbbf9b47d0c0ff7d6ebcb78
Successfully built smart-open


In [24]:
import warnings
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
import gensim
from gensim.models import Word2Vec

warnings.filterwarnings("ignore")

1. Создайте мешок слов с помощью sklearn.feature_extraction.text.CountVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
Исключим стоп-слова с помощью stop_words='english'.
Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью CountVectorizer.get_feature_names().

In [4]:
combine_df = pd.read_csv('./combine_df.csv')
combine_df.head()

id  label                                              tweet  \
0   1    0.0   @user when a father is dysfunctional and is s...   
1   2    0.0  @user @user thanks for #lyft credit i can't us...   
2   3    0.0                                bihday your majesty   
3   4    0.0  #model   i love u take with u all the time in ...   
4   5    0.0             factsguide: society now    #motivation   

                                         clean_tweet  \
0  when fatheare is dysfyounctional and is so sel...   
1  thanks foare lyft careedit cannot youse cayous...   
2                            bihday yoyouare majesty   
3  model love you take with you all the time in y...   
4           factsgareinyouide society now motivation   

                                         tweet_token  \
0  ['when', 'fatheare', 'is', 'dysfyounctional', ...   
1  ['thanks', 'foare', 'lyft', 'careedit', 'can',...   
2                  ['bihday', 'yoyouare', 'majesty']   
3  ['model', 'love', 'you', 'take', 'with', 'you'...   
4  ['factsgareinyouide', 'society', 'now', 'motiv...   

                                tweet_token_filtered  \
0  ['fatheare', 'dysfyounctional', 'selfish', 'da...   
1  ['thanks', 'foare', 'lyft', 'careedit', 'youse...   
2                  ['bihday', 'yoyouare', 'majesty']   
3        ['model', 'love', 'take', 'time', 'youare']   
4     ['factsgareinyouide', 'society', 'motivation']   

                                       tweet_stemmed  \
0  ['fathear', 'dysfyounct', 'selfish', 'dareagar...   
1  ['thank', 'foar', 'lyft', 'careedit', 'yous', ...   
2                   ['bihday', 'yoyouar', 'majesti']   
3         ['model', 'love', 'take', 'time', 'youar']   
4           ['factsgareinyouid', 'societi', 'motiv']   

                                    tweet_lemmatized  
0  ['fatheare', 'dysfyounctional', 'selfish', 'da...  
1  ['thank', 'foare', 'lyft', 'careedit', 'youse'...  
2                  ['bihday', 'yoyouare', 'majesty']  
3        ['model', 'love', 'take', 'time', 'youare']  
4     ['factsgareinyouide', 'society', 'motivation']

In [7]:
count_vectorizer = CountVectorizer(ngram_range=(1, 1), max_df=0.9, max_features = 1000, stop_words='english')

In [8]:
# Создаем the Bag-of-Words модель
bag_of_words_stemmed = count_vectorizer.fit_transform(combine_df['tweet_stemmed'])

In [9]:
# Отобразим Bag-of-Words модель как DataFrame
feature_names = count_vectorizer.get_feature_names()
pd.DataFrame(bag_of_words_stemmed.toarray(), columns = feature_names)

aare  aareareiv  aareoyound  abl  aboyout  absolyout  accept  \
0         0          0           0    0        0          0       0   
1         0          0           0    0        0          0       0   
2         0          0           0    0        0          0       0   
3         0          0           0    0        0          0       0   
4         0          0           0    0        0          0       0   
...     ...        ...         ...  ...      ...        ...     ...   
49154     0          0           0    0        0          0       0   
49155     0          0           0    0        0          0       0   
49156     0          0           0    0        0          0       0   
49157     0          0           0    0        0          0       0   
49158     0          0           0    0        0          0       0   

       accoyount  act  action  ...  yoyouarear  yoyouareself  yoyoultyouare  \
0              0    0       0  ...           0             0              0   
1              0    0       0  ...           0             0              0   
2              0    0       0  ...           0             0              0   
3              0    0       0  ...           0             0              0   
4              0    0       0  ...           0             0              0   
...          ...  ...     ...  ...         ...           ...            ...   
49154          0    0       0  ...           0             0              0   
49155          0    0       0  ...           0             0              0   
49156          0    0       0  ...           0             0              0   
49157          0    0       0  ...           0             0              1   
49158          0    0       0  ...           0             0              0   

       yoyoungarein  yoyoup  yoyous  yoyoustomear  yoyout  yoyoutyoub  yyoummi  
0                 0       0       0             0       0           0        0  
1                 0       0       0             0       0           0        0  
2                 0       0       0             0       0           0        0  
3                 0       0       0             0       0           0        0  
4                 0       0       0             0       0           0        0  
...             ...     ...     ...           ...     ...         ...      ...  
49154             0       0       0             0       0           0        0  
49155             0       0       0             0       0           0        0  
49156             0       0       0             0       0           0        0  
49157             0       0       0             0       0           0        0  
49158             0       0       0             0       0           0        0  

[49159 rows x 1000 columns]

In [10]:
bag_of_words_lemmatized = count_vectorizer.fit_transform(combine_df['tweet_lemmatized'])

In [11]:
feature_names = count_vectorizer.get_feature_names()
pd.DataFrame(bag_of_words_lemmatized.toarray(), columns = feature_names)

aareareived  aaree  aareoyound  able  aboyout  absolyoutely  accept  \
0                0      0           0     0        0             0       0   
1                0      0           0     0        0             0       0   
2                0      0           0     0        0             0       0   
3                0      0           0     0        0             0       0   
4                0      0           0     0        0             0       0   
...            ...    ...         ...   ...      ...           ...     ...   
49154            0      0           0     0        0             0       0   
49155            0      0           0     0        0             0       0   
49156            0      0           0     0        0             0       0   
49157            0      0           0     0        0             0       0   
49158            0      0           0     0        0             0       0   

       accoyount  act  action  ...  yoyoultyouaree  yoyoungarein  yoyoup  \
0              0    0       0  ...               0             0       0   
1              0    0       0  ...               0             0       0   
2              0    0       0  ...               0             0       0   
3              0    0       0  ...               0             0       0   
4              0    0       0  ...               0             0       0   
...          ...  ...     ...  ...             ...           ...     ...   
49154          0    0       0  ...               0             0       0   
49155          0    0       0  ...               0             0       0   
49156          0    0       0  ...               0             0       0   
49157          0    0       0  ...               1             0       0   
49158          0    0       0  ...               0             0       0   

       yoyous  yoyoustomeare  yoyoustomeares  yoyout  yoyoute  yoyoutyoube  \
0           0              0               0       0        0            0   
1           0              0               0       0        0            0   
2           0              0               0       0        0            0   
3           0              0               0       0        0            0   
4           0              0               0       0        0            0   
...       ...            ...             ...     ...      ...          ...   
49154       0              0               0       0        0            0   
49155       0              0               0       0        0            0   
49156       0              0               0       0        0            0   
49157       0              0               0       0        0            0   
49158       0              0               0       0        0            0   

       yyoummy  
0            0  
1            0  
2            0  
3            0  
4            0  
...        ...  
49154        0  
49155        0  
49156        0  
49157        0  
49158        0  

[49159 rows x 1000 columns]

2. Создайте мешок слов с помощью sklearn.feature_extraction.text.TfidfVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
Исключим стоп-слова с помощью stop_words='english'.
Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью TfidfVectorizer.get_feature_names().

In [13]:
tfidf_vectorizer = TfidfVectorizer(ngram_range=(1, 1), max_df=0.9, max_features = 1000, stop_words='english')

In [14]:
values_stemmed = tfidf_vectorizer.fit_transform(combine_df['tweet_stemmed'])

In [15]:
feature_names = tfidf_vectorizer.get_feature_names()
pd.DataFrame(values_stemmed.toarray(), columns = feature_names)

aare  aareareiv  aareoyound  abl  aboyout  absolyout  accept  \
0       0.0        0.0         0.0  0.0      0.0        0.0     0.0   
1       0.0        0.0         0.0  0.0      0.0        0.0     0.0   
2       0.0        0.0         0.0  0.0      0.0        0.0     0.0   
3       0.0        0.0         0.0  0.0      0.0        0.0     0.0   
4       0.0        0.0         0.0  0.0      0.0        0.0     0.0   
...     ...        ...         ...  ...      ...        ...     ...   
49154   0.0        0.0         0.0  0.0      0.0        0.0     0.0   
49155   0.0        0.0         0.0  0.0      0.0        0.0     0.0   
49156   0.0        0.0         0.0  0.0      0.0        0.0     0.0   
49157   0.0        0.0         0.0  0.0      0.0        0.0     0.0   
49158   0.0        0.0         0.0  0.0      0.0        0.0     0.0   

       accoyount  act  action  ...  yoyouarear  yoyouareself  yoyoultyouare  \
0            0.0  0.0     0.0  ...         0.0           0.0       0.000000   
1            0.0  0.0     0.0  ...         0.0           0.0       0.000000   
2            0.0  0.0     0.0  ...         0.0           0.0       0.000000   
3            0.0  0.0     0.0  ...         0.0           0.0       0.000000   
4            0.0  0.0     0.0  ...         0.0           0.0       0.000000   
...          ...  ...     ...  ...         ...           ...            ...   
49154        0.0  0.0     0.0  ...         0.0           0.0       0.000000   
49155        0.0  0.0     0.0  ...         0.0           0.0       0.000000   
49156        0.0  0.0     0.0  ...         0.0           0.0       0.000000   
49157        0.0  0.0     0.0  ...         0.0           0.0       0.279313   
49158        0.0  0.0     0.0  ...         0.0           0.0       0.000000   

       yoyoungarein  yoyoup  yoyous  yoyoustomear  yoyout  yoyoutyoub  yyoummi  
0               0.0     0.0     0.0           0.0     0.0         0.0      0.0  
1               0.0     0.0     0.0           0.0     0.0         0.0      0.0  
2               0.0     0.0     0.0           0.0     0.0         0.0      0.0  
3               0.0     0.0     0.0           0.0     0.0         0.0      0.0  
4               0.0     0.0     0.0           0.0     0.0         0.0      0.0  
...             ...     ...     ...           ...     ...         ...      ...  
49154           0.0     0.0     0.0           0.0     0.0         0.0      0.0  
49155           0.0     0.0     0.0           0.0     0.0         0.0      0.0  
49156           0.0     0.0     0.0           0.0     0.0         0.0      0.0  
49157           0.0     0.0     0.0           0.0     0.0         0.0      0.0  
49158           0.0     0.0     0.0           0.0     0.0         0.0      0.0  

[49159 rows x 1000 columns]

In [18]:
values_lemmatized = tfidf_vectorizer.fit_transform(combine_df['tweet_lemmatized'])

In [19]:
feature_names = tfidf_vectorizer.get_feature_names()
pd.DataFrame(values_lemmatized.toarray(), columns = feature_names)

aareareived  aaree  aareoyound  able  aboyout  absolyoutely  accept  \
0              0.0    0.0         0.0   0.0      0.0           0.0     0.0   
1              0.0    0.0         0.0   0.0      0.0           0.0     0.0   
2              0.0    0.0         0.0   0.0      0.0           0.0     0.0   
3              0.0    0.0         0.0   0.0      0.0           0.0     0.0   
4              0.0    0.0         0.0   0.0      0.0           0.0     0.0   
...            ...    ...         ...   ...      ...           ...     ...   
49154          0.0    0.0         0.0   0.0      0.0           0.0     0.0   
49155          0.0    0.0         0.0   0.0      0.0           0.0     0.0   
49156          0.0    0.0         0.0   0.0      0.0           0.0     0.0   
49157          0.0    0.0         0.0   0.0      0.0           0.0     0.0   
49158          0.0    0.0         0.0   0.0      0.0           0.0     0.0   

       accoyount  act  action  ...  yoyoultyouaree  yoyoungarein  yoyoup  \
0            0.0  0.0     0.0  ...         0.00000           0.0     0.0   
1            0.0  0.0     0.0  ...         0.00000           0.0     0.0   
2            0.0  0.0     0.0  ...         0.00000           0.0     0.0   
3            0.0  0.0     0.0  ...         0.00000           0.0     0.0   
4            0.0  0.0     0.0  ...         0.00000           0.0     0.0   
...          ...  ...     ...  ...             ...           ...     ...   
49154        0.0  0.0     0.0  ...         0.00000           0.0     0.0   
49155        0.0  0.0     0.0  ...         0.00000           0.0     0.0   
49156        0.0  0.0     0.0  ...         0.00000           0.0     0.0   
49157        0.0  0.0     0.0  ...         0.27931           0.0     0.0   
49158        0.0  0.0     0.0  ...         0.00000           0.0     0.0   

       yoyous  yoyoustomeare  yoyoustomeares  yoyout  yoyoute  yoyoutyoube  \
0         0.0            0.0             0.0     0.0      0.0          0.0   
1         0.0            0.0             0.0     0.0      0.0          0.0   
2         0.0            0.0             0.0     0.0      0.0          0.0   
3         0.0            0.0             0.0     0.0      0.0          0.0   
4         0.0            0.0             0.0     0.0      0.0          0.0   
...       ...            ...             ...     ...      ...          ...   
49154     0.0            0.0             0.0     0.0      0.0          0.0   
49155     0.0            0.0             0.0     0.0      0.0          0.0   
49156     0.0            0.0             0.0     0.0      0.0          0.0   
49157     0.0            0.0             0.0     0.0      0.0          0.0   
49158     0.0            0.0             0.0     0.0      0.0          0.0   

       yyoummy  
0          0.0  
1          0.0  
2          0.0  
3          0.0  
4          0.0  
...        ...  
49154      0.0  
49155      0.0  
49156      0.0  
49157      0.0  
49158      0.0  

[49159 rows x 1000 columns]

3. Натренируем gensim.models.Word2Vec модель на наших данных.
Тренировать будем на токенизированных твитах combine_df['tweet_token']
Установим следующие параметры: size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34.
Используем функцию train() с параметром total_examples равным длине combine_df['tweet_token'], количество epochs установим 20.

In [30]:
modelW2V = Word2Vec(sentences=combine_df['tweet_token'],  size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34)

In [33]:
modelW2V.train(combine_df['tweet_token'], total_examples=len(combine_df['tweet_token']), epochs=20)

(22459195, 126605280)

4. Давайте немного потестируем нашу модель Word2Vec и посмотрим, как она работает. Мы зададим слово positive = "dinner", и модель вытащит из корпуса наиболее похожие слова c помощью функции most_similar. То же самое попробуем со словом "trump".

In [37]:
modelW2V.most_similar(positive = "dinner")

KeyError: "word 'dinner' not in vocabulary"

5. Из приведенных выше примеров мы видим, что наша модель word2vec хорошо справляется с поиском наиболее похожих слов для данного слова. Но как она это делает? Она изучила векторы для каждого уникального слова наших данных и использует косинусное сходство, чтобы найти наиболее похожие векторы (слова).
Давайте проверим векторное представление любого слова из нашего корпуса, например "food".

6. Поскольку наши данные содержат твиты, а не только слова, нам придется придумать способ использовать векторы слов из модели word2vec для создания векторного представления всего твита. Существует простое решение этой проблемы, мы можем просто взять среднее значение всех векторов слов, присутствующих в твите. Длина результирующего вектора будет одинаковой, то есть 200. Мы повторим тот же процесс для всех твитов в наших данных и получим их векторы. Теперь у нас есть 200 функций word2vec для наших данных.
Необходимо создать вектор для каждого твита, взяв среднее значение векторов слов, присутствующих в твите. В цикле сделать:  vec += model_w2v[word].reshape((1, size))
и поделить финальный вектор на количество слов в твите.
На выходе должен получиться wordvec_df.shape = (49159, 200)